src/main/scala/is/hail/expr/ir/Primitives.scala

```diff
@@ -0,0 +1,57 @@
+package is.hail.expr.ir
+
+import is.hail.utils._
+import is.hail.annotations.MemoryBuffer
+import is.hail.asm4s._
+import is.hail.expr.{TInt32, TInt64, TArray, TContainer, TStruct, TFloat32, TFloat64, TBoolean, Type}
+import is.hail.annotations.StagedRegionValueBuilder
+
+import scala.collection.mutable
+
+object Primitives {
+  private case class Primitive(name: String, f: (Array[Type]) => (Type, Array[Code[_]] => Code[_]))
+  private val primitives: mutable.Map[String, Primitive] = mutable.HashMap()
+  private def numeric(
+    name: String,
+    i: (Code[Int], Code[Int]) => Code[Int],
+    l: (Code[Long], Code[Long]) => Code[Long],
+    f: (Code[Float], Code[Float]) => Code[Float],
+    d: (Code[Double], Code[Double]) => Code[Double]
+  ): Primitive = Primitive(name,
+    { case Array(TInt32, TInt32)     => TInt32   -> { case Array(x, y) => i(x.asInstanceOf[Code[Int]], y.asInstanceOf[Code[Int]]) }
+      case Array(TInt64, TInt64)     => TInt64   -> { case Array(x, y) => l(x.asInstanceOf[Code[Long]], y.asInstanceOf[Code[Long]]) }
+      case Array(TFloat32, TFloat32) => TFloat32 -> { case Array(x, y) => f(x.asInstanceOf[Code[Float]], y.asInstanceOf[Code[Float]]) }
+      case Array(TFloat64, TFloat64) => TFloat64 -> { case Array(x, y) => d(x.asInstanceOf[Code[Double]], y.asInstanceOf[Code[Double]]) }
+
+      case Array(TInt32, TInt64)     => TInt64   -> { case Array(x, y) => l(x.asInstanceOf[Code[Int]].toL, y.asInstanceOf[Code[Long]]) }
```

Comment:
I'd prefer if conversions are explicit in the IR.

Reply:
Sure. I put this in so I could have some non-trivial tests. I could go all the way and hook it up to the FunctionRegistry if you'd prefer.

Reply:
I'd request a `Cast(t: Type, arg: IR)` that casts `arg` to type `t`.  Then casts are explicit (and can be optimized so), the conversions in the function registry can produce the cast nodes as needed, and you can still code up non-trivial examples.

Just to be clear, I was just talking about the arithmetic operations with differing types.  Things like `Int32 + Int32` should stay.

Reply:
Cast is now in the IR and used explicitly in the tests, this `Primitives` object no longer does casting.